src/AgentImplementation.sol

```diff
@@ -298,36 +293,29 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
     }
 
     function _chargeFeeByMsgValue() internal {
-        if (msg.value == 0) return;
-
-        // TODO: implement new msg.value fee charge mechanism
+        if (msg.value == 0) {
+            return;
+        } else {
+            uint256 feeRate = IRouter(router).feeRate();
+            address feeCollector = IRouter(router).feeCollector();
+            IParam.Fee memory fee = FeeLogic.getFee(_NATIVE, msg.value, feeRate, _NATIVE_FEE_META_DATA);
+            fee.charge(feeCollector);
+        }
     }
 
-    function _chargeFee(IParam.Fee[] memory fees, bool chargeFromUser) internal {
+    function _chargeFee(IParam.Fee[] memory fees, bool chargeFromAgent) internal {
         uint256 length = fees.length;
         if (length == 0) return;
 
         address feeCollector = IRouter(router).feeCollector();
         for (uint256 i; i < length; ) {
-            uint256 amount = fees[i].amount;
-            if (amount == 0) {
-                unchecked {
-                    ++i;
-                }
-                continue;
-            }
-
             address token = fees[i].token;
-            if (token == _NATIVE) {
-                payable(feeCollector).sendValue(amount);
-            } else if (chargeFromUser) {
-                address user = IRouter(router).currentUser();
-                IAllowanceTransfer(permit2).transferFrom(user, feeCollector, amount.toUint160(), token);
+            if (token == _NATIVE || chargeFromAgent) {
+                fees[i].charge(feeCollector);
             } else {
-                IERC20(token).safeTransfer(feeCollector, amount);
+                address user = IRouter(router).currentUser();
+                fees[i].chargeFrom(feeCollector, user, permit2);
```

Comment:
Can we use `IRouter(router).currentUser()` directly in `chargeFrom()` to save some gas?

Reply:
I think it should be move to L311 to avoid repeatedly calling.

Reply:
In most cases, the token is set to `NATIVE`. Moving it to L311 would result in the `currentUser` being called unnecessarily in the majority of scenarios. Additionally, the `else` condition is usually used for delegation or signature cases, where the fee token is typically calculated off-chain as a single token. Hence, I'd prefer to maintain the current implementation.

Reply:
That's a fair point also. I prefer to stay with the current version. Can call in the function directly though i don't think it makes much different.